<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Changepoint Additive Models for Time Series Omics Data • cpam</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Changepoint Additive Models for Time Series Omics Data">
<meta name="description" content="Provides a comprehensive framework for time series omics analysis, integrating changepoint detection, smooth and shape-constrained trends, and uncertainty quantification. It supports gene- and transcript-level inferences, p-value aggregation for improved power, and both case-only and case-control designs. It includes an interactive Shiny interface.">
<meta property="og:description" content="Provides a comprehensive framework for time series omics analysis, integrating changepoint detection, smooth and shape-constrained trends, and uncertainty quantification. It supports gene- and transcript-level inferences, p-value aggregation for improved power, and both case-only and case-control designs. It includes an interactive Shiny interface.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">cpam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="cpam-changepoint-additive-models">cpam <span style="font-weight: normal">(<strong>c</strong>hange<strong>p</strong>oint <strong>a</strong>dditive <strong>m</strong>odels)</span>
<a class="anchor" aria-label="anchor" href="#cpam-changepoint-additive-models"></a>
</h2>
<!-- badges: start -->

<div class="section level3">
<h3 id="an-r-package-for-omics-time-series-analysis">An R package for omics time series analysis.<a class="anchor" aria-label="anchor" href="#an-r-package-for-omics-time-series-analysis"></a>
</h3>
<p>Read the methods paper <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">here</a></p>
<p><img src="reference/figures/example_trends.png" width="800" height="600"></p>
<p>Application of <strong>cpam</strong> to RNA-seq time series of <em>Arabidopsis</em> plants treated with excess-light.</p>
</div>
</div>
<div class="section level2">
<h2 id="key-features">Key features<a class="anchor" aria-label="anchor" href="#key-features"></a>
</h2>
<ul>
<li>
<strong>Changepoint detection</strong>: Identify sharp transitions in expression.</li>
<li>
<strong>smooth trends</strong>: Model expression as a smooth function of time.</li>
<li>
<strong>Shape-constrained trends</strong>: Cluster targets into biologically meaningful temporal shape classes.</li>
<li>
<strong>Quantification uncertainty</strong>: Account for uncertainty in expression estimates.</li>
<li>
<strong>Transcript-level analysis</strong>
<ul>
<li>Perform gene- or transcript-level inferences.</li>
<li>Aggregate <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values at the gene level for improved power.</li>
</ul>
</li>
<li>
<strong>Case-only or case-control time series</strong>: Analyse time series data with or without controls.</li>
<li>
<strong>User-friendly</strong>: Sensible defaults and an interactive <em>shiny</em> interface.</li>
</ul>
<p>Our new package <strong>cpam</strong> provides a comprehensive framework for analysing time series omics data. The method uses modern statistical approaches while remaining user-friendly, through sensible defaults and an interactive interface. Researchers can directly address key questions in time series analysis—when changes occur, what patterns they follow, and how responses are related. While we have focused on transcriptomics, the framework is applicable to other high-dimensional time series measurements, making it a valuable addition to the omics analysis toolkit.</p>
<p>This is the initial development version of <strong>cpam</strong> and we are actively seeking community feedback. If you encounter issues or have suggestions for improvements, please open an <a href="https://github.com/l-a-yates/cpam/issues" class="external-link">issue</a>. We welcome questions and discussion about using <strong>cpam</strong> for your research through <a href="https://github.com/l-a-yates/cpam/discussions/" class="external-link">Discussions</a>. Our goal is to work with users to make <strong>cpam</strong> a robust and valuable tool for time series omics analysis. We can also be contacted via the email addresses listed in our paper <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Installation code here</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"l-a-yates/cpam"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="step-1-load-the-package">Step 1: Load the package<a class="anchor" aria-label="anchor" href="#step-1-load-the-package"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://l-a-yates.github.io/cpam/">cpam</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-create-a-tibble-for-the-experimental-design">Step 2: Create a tibble for the experimental design.<a class="anchor" aria-label="anchor" href="#step-2-create-a-tibble-for-the-experimental-design"></a>
</h3>
<p>In this <em>Arabidopsis thaliana</em> time series example, we used the software <a href="https://doi.org/10.1038/nbt.3519" class="external-link">kallisto</a> to generate counts from RNA-seq data. To load the counts, we provide the file path for each kallisto output file (alternatively you can provide the counts directly as count matrix, or use other quantification software)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_design_path</span></span>
<span><span class="co">#&gt; # A tibble: 50 × 4</span></span>
<span><span class="co">#&gt;    sample  time path                                condition</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;                               &lt;chr&gt;    </span></span>
<span><span class="co">#&gt;  1 JHSS01     0 output/kallisto/JHSS01/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  2 JHSS02     0 output/kallisto/JHSS02/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  3 JHSS03     0 output/kallisto/JHSS03/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  4 JHSS04     0 output/kallisto/JHSS04/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  5 JHSS05     0 output/kallisto/JHSS05/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  6 JHSS06     5 output/kallisto/JHSS06/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  7 JHSS07     5 output/kallisto/JHSS07/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  8 JHSS08     5 output/kallisto/JHSS08/abundance.h5 treatment</span></span>
<span><span class="co">#&gt;  9 JHSS09     5 output/kallisto/JHSS09/abundance.h5 treatment</span></span>
<span><span class="co">#&gt; 10 JHSS10     5 output/kallisto/JHSS10/abundance.h5 treatment</span></span>
<span><span class="co">#&gt; # ℹ 40 more rows</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-obtain-a-table-with-the-transcript-to-gene-mapping">Step 3: Obtain a table with the transcript-to-gene mapping<a class="anchor" aria-label="anchor" href="#step-3-obtain-a-table-with-the-transcript-to-gene-mapping"></a>
</h3>
<p>N.B. This is not needed if your counts are aggregated at the gene level, but transcript-level analysis with aggregation of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values to the gene level is recommended. E.g., for <em>Arabidopsis thaliana</em>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">t2g_arabidopsis</span></span>
<span><span class="co">#&gt; # A tibble: 50 × 2</span></span>
<span><span class="co">#&gt;    target_id   gene_id  </span></span>
<span><span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;    </span></span>
<span><span class="co">#&gt;  1 AT1G01010.1 AT1G01010</span></span>
<span><span class="co">#&gt;  2 AT1G01020.2 AT1G01020</span></span>
<span><span class="co">#&gt;  3 AT1G01020.6 AT1G01020</span></span>
<span><span class="co">#&gt;  4 AT1G01020.1 AT1G01020</span></span>
<span><span class="co">#&gt;  5 AT1G01020.4 AT1G01020</span></span>
<span><span class="co">#&gt;  6 AT1G01020.5 AT1G01020</span></span>
<span><span class="co">#&gt;  7 AT1G01020.3 AT1G01020</span></span>
<span><span class="co">#&gt;  8 AT1G03987.1 AT1G03987</span></span>
<span><span class="co">#&gt;  9 AT1G01030.2 AT1G01030</span></span>
<span><span class="co">#&gt; 10 AT1G01030.1 AT1G01030</span></span>
<span><span class="co">#&gt; # ℹ 40 more rows</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-4-run-cpam">Step 4: Run <strong>cpam</strong>
<a class="anchor" aria-label="anchor" href="#step-4-run-cpam"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/prepare_cpam.html">prepare_cpam</a></span><span class="op">(</span>exp_design <span class="op">=</span> <span class="va">exp_design_path</span>,</span>
<span>                      count_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      t2g <span class="op">=</span> <span class="va">t2g_arabidopsis</span>,</span>
<span>                      model <span class="op">=</span> <span class="st">"case-only"</span>,</span>
<span>                      import_type <span class="op">=</span> <span class="st">"kallisto"</span>,</span>
<span>                      num_cores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_p_values.html">compute_p_values</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> </span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/estimate_changepoint.html">estimate_changepoint</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> </span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/select_shape.html">select_shape</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-5-visualise-the-results">Step 5: Visualise the results<a class="anchor" aria-label="anchor" href="#step-5-visualise-the-results"></a>
</h3>
<p>Load the shiny app for an interactive visualisation of the results:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="reference/visualize.html">visualise</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="co"># not shown in vignette</span></span></code></pre></div>
<p>Or plot one gene at a time:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="reference/plot_cpam.html">plot_cpam</a></span><span class="op">(</span><span class="va">cpo</span>, gene_id <span class="op">=</span> <span class="st">"AT3G23280"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/example_gene_plot.png" width="650" height="550"></p>
<p>Isoform 1 (AT3G23280.1) has a changepoint at 67.5 min and has a monotonic increasing concave (micv) shape. Isoform 2 (AT3G23280.2) has no changepoint and has an unconstrained thin-plate (tp) shape. <br><br></p>
<p>We can generate a results table which has <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values, shapes, log-fold changes and counts with many optimal filters (see tutorials):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 15,279 × 25</span></span>
<span><span class="co">#&gt;    target_id   gene_id     p    cp shape lfc.0 lfc.5 lfc.10 lfc.20 lfc.30 lfc.45</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 AT1G01910.1 AT1G01…     0     0 micv      0 1.01   1.70   2.38   2.60   2.73 </span></span>
<span><span class="co">#&gt;  2 AT1G01910.2 AT1G01…     0    10 cv        0 0      0      0.553  0.775  0.790</span></span>
<span><span class="co">#&gt;  3 AT1G01910.5 AT1G01…     0    10 cx        0 0      0     -3.20  -4.57  -4.82 </span></span>
<span><span class="co">#&gt;  4 AT1G02610.1 AT1G02…     0    45 mdcx      0 0      0      0      0      0    </span></span>
<span><span class="co">#&gt;  5 AT1G02610.2 AT1G02…     0    10 cx        0 0      0     -0.645 -1.16  -1.71 </span></span>
<span><span class="co">#&gt;  6 AT1G02610.3 AT1G02…     0    10 mdcx      0 0      0     -1.48  -2.11  -2.25 </span></span>
<span><span class="co">#&gt;  7 AT1G04080.1 AT1G04…     0    10 cv        0 0      0      2.75   3.85   3.97 </span></span>
<span><span class="co">#&gt;  8 AT1G04080.2 AT1G04…     0    45 micv      0 0      0      0      0      0    </span></span>
<span><span class="co">#&gt;  9 AT1G04080.3 AT1G04…     0     0 micv      0 0.268  0.445  0.603  0.638  0.656</span></span>
<span><span class="co">#&gt; 10 AT1G04080.5 AT1G04…     0    10 cx        0 0      0     -2.17  -3.04  -3.10 </span></span>
<span><span class="co">#&gt; # ℹ 15,269 more rows</span></span>
<span><span class="co">#&gt; # ℹ 14 more variables: lfc.60 &lt;dbl&gt;, lfc.90 &lt;dbl&gt;, lfc.180 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   lfc.240 &lt;dbl&gt;, counts.0 &lt;dbl&gt;, counts.5 &lt;dbl&gt;, counts.10 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   counts.20 &lt;dbl&gt;, counts.30 &lt;dbl&gt;, counts.45 &lt;dbl&gt;, counts.60 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   counts.90 &lt;dbl&gt;, counts.180 &lt;dbl&gt;, counts.240 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="tutorials">Tutorials<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h3>
<p>For a quick-to-run introductory example, we have provided a small simulated data set as part of the package.</p>
<ul>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/example.html" class="external-link">Introductory Example</a></li>
</ul>
<p>The following two tutorials use real-world data to demonstrate the capabilities of the <code>cpam</code> package. In addition, they provide code to reproduce the results for the case studies presented in the <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">manuscript</a> accompanying the <code>cpam</code> package.</p>
<ul>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/crisp.html" class="external-link">Arabidopsis Case Study</a></li>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/torre.html" class="external-link">Human Embryo Case Study</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>This work was supported by the Australian Research Council, Centre of Excellence for Plant Success in Nature and Agriculture (CE200100015).</p>
<p><img src="reference/figures/plant_success_logo_horizontal.jpg" width="100%" height="auto" alt="Plant Success Logo"></p>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing cpam</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Luke Yates <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0000-0002-1685-3169" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Michael Charleston <br><small class="roles"> Author </small>  </li>
<li>Jazmine Humphreys <br><small class="roles"> Author </small>  </li>
<li>Steven Smith <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/l-a-yates/cpam/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/l-a-yates/cpam/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luke Yates, Michael Charleston, Jazmine Humphreys, Steven Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
