---
title: false
output: 
  github_document:
    toc: false
    toc_depth: 2
    html_preview: false
knit: (function(inputFile, encoding) {
    rmarkdown::render(inputFile, encoding = encoding, output_dir = ".")
  })
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```



## cpam <span style="font-weight: normal">(**c**hange**p**oint **a**dditive **m**odels)</span>
An R package for omics time series analysis.

Read the full paper [here](https://www.biorxiv.org/)

![](assets/example_trends.png){width=800 height=600}
Application of cpam to RNA-seq time series of *Arabidopsis* plants treated with excess-light. 

## Key features
 - **Changepoint detection**: Identify sharp transitions in expression.
 - **smooth trends**: Model expression as a smooth function of time.
 - **Shape-constrained trends**: Cluster genes or isoforms into biologically meaningful temporal shape classes.
 - **Quantification uncertainty**: Account for uncertainty in expression estimates.
 - **Transcript-level analysis**
    - Perform gene- or transcript-level inferences. 
    - Aggregate $p$-values at the gene level for improved power.
 - **Case-only or case-control time series**: Analyse time series data with or without controls.
 - **User-friendly**: Sensible defaults and an interactive *shiny* interface.

Our new package **cpam** provides a comprehensive framework for analysing time series omics data that combines statistical rigour with practical utility. The method leverages modern statistical approaches while remaining user-friendly, through sensible defaults and an interactive interface. 
Researchers can directly address key questions in time series analysis---when changes occur, what patterns they follow, and how responses are related. While we have focused on transcriptomics, the framework is applicable to other high-dimensional time series measurements, making it a valuable addition to the omics analysis toolkit.


This is the initial release of **cpam** and we are actively seeking community feedback. If you encounter issues or have suggestions for improvements, please open an [issue](https://github.com/username/cpam/issues). We welcome questions and discussion about using **cpam** for your research through [Discussions](https://github.com/username/cpam/discussions). Our goal is to work with users to make **cpam** a robust and valuable tool for time series omics analysis. We can also be contacted via the email addresses listed in the paper [here](https://www.biorxiv.org/). 

## Installation
```{r, eval = FALSE}
# Installation code here
remotes::install_github("l-a-yates/cpam")
```

## Usage

### Step 1: Load the package
```{r, echo = TRUE}
library(cpam)
```
### Step 2: Create a tibble (or data frame) for the experimental design. In this example, we are using kallisto-generated counts, so we provide the file path for each sample (otherwise you can provide the counts directly as count matrix)
```{r, echo = TRUE}
library(cpam)

exp_design
```
### Step 3: You will need source a table with the transcript-to-gene mapping, unless your counts are aggregated at the gene level (transcript-level analysis with aggregation of $p$-values to the gene level is recommened). E.g., for *Arabidopsis thaliana*:
```{r}
t2g
```
### Step 4: Run **cpam** 
```{r, eval = FALSE}
  cpo <- prepare_cpam(exp_design = exp_design,
                      count_matrix = NULL,
                      t2g = t2g,
                      import_type = "kallisto",
                      num_cores = 5) # about 1 min
  cpo <- compute_p_values(cpo) # 30 secs to 5 min
  cpo <- estimate_changepoint(cpo) # 1-10 min
  cpo <- select_shape(cpo) # 1-20 min

```
### Step 5: Visualise the results

Load the shiny app for an interactive visualisation of the results:
```{r, eval = FALSE}
  visualise(cpo)
```
Or plot one gene at a time:
```{r, eval = FALSE}
  plot_cpam(cpo, gene_id = "AT3G23280")
```
Generate a results table:
```{r, eval = FALSE}
  results(cpo)
```


