<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot fitted changepoint additive models — plot_cpam • cpam</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot fitted changepoint additive models — plot_cpam"><meta name="description" content="Plot fitted changepoint additive models"><meta property="og:description" content="Plot fitted changepoint additive models"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cpam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cpam.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/l-a-yates/cpam/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Plot fitted changepoint additive models</h1>
      <small class="dont-index">Source: <a href="https://github.com/l-a-yates/cpam/blob/main/R/plot-gene.R" class="external-link"><code>R/plot-gene.R</code></a></small>
      <div class="d-none name"><code>plot_cpam.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Plot fitted changepoint additive models</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_cpam</span><span class="op">(</span></span>
<span>  <span class="va">cpo</span>,</span>
<span>  gene_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  target_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cp_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cp_1se"</span>, <span class="st">"cp_min"</span><span class="op">)</span>,</span>
<span>  shape_type <span class="op">=</span> <span class="st">"shape1"</span>,</span>
<span>  bs <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  cp_fix <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  facet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_fit <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_data <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_fit_ci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_data_ci <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ci_prob <span class="op">=</span> <span class="st">"se"</span>,</span>
<span>  remove_null <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  null_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  null_threshold_adj <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  k_mult <span class="op">=</span> <span class="fl">1.2</span>,</span>
<span>  return_fits_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>  common_y_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaled <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cpo">cpo<a class="anchor" aria-label="anchor" href="#arg-cpo"></a></dt>
<dd><p>A cpam object containing count data, model fits, and optional changepoint/shape estimates</p></dd>


<dt id="arg-gene-id">gene_id<a class="anchor" aria-label="anchor" href="#arg-gene-id"></a></dt>
<dd><p>character; gene_id (mutually exclusive with target_id)</p></dd>


<dt id="arg-target-id">target_id<a class="anchor" aria-label="anchor" href="#arg-target-id"></a></dt>
<dd><p>character; target_id (mutually exclusive with gene_id)</p></dd>


<dt id="arg-cp-type">cp_type<a class="anchor" aria-label="anchor" href="#arg-cp-type"></a></dt>
<dd><p>One of "cp_1se" or "cp_min"; rule for selecting changepoint from fitted models.
See <code><a href="estimate_changepoint.html">estimate_changepoint()</a></code> for details.</p></dd>


<dt id="arg-shape-type">shape_type<a class="anchor" aria-label="anchor" href="#arg-shape-type"></a></dt>
<dd><p>One of "shape1" or "shape2"; which set of fitted shape patterns to use.
See <code><a href="select_shape.html">select_shape()</a></code> for details.</p></dd>


<dt id="arg-bs">bs<a class="anchor" aria-label="anchor" href="#arg-bs"></a></dt>
<dd><p>Shape pattern to fit ("null", "lin", "ilin", "dlin", or from cpo$bss).
Use "auto" (default) to use estimated shapes as per <code>shape_type</code>.</p></dd>


<dt id="arg-cp-fix">cp_fix<a class="anchor" aria-label="anchor" href="#arg-cp-fix"></a></dt>
<dd><p>Numeric; fixed changepoint time. Set to -1 (default) to use estimated changepoints</p></dd>


<dt id="arg-facet">facet<a class="anchor" aria-label="anchor" href="#arg-facet"></a></dt>
<dd><p>Logical; for multiple transcripts, plot in separate facets?</p></dd>


<dt id="arg-sp">sp<a class="anchor" aria-label="anchor" href="#arg-sp"></a></dt>
<dd><p>numerical; set the smooth parameter. NULL (default) for automatic selection</p></dd>


<dt id="arg-show-fit">show_fit<a class="anchor" aria-label="anchor" href="#arg-show-fit"></a></dt>
<dd><p>logical; show the fitted trend?</p></dd>


<dt id="arg-show-data">show_data<a class="anchor" aria-label="anchor" href="#arg-show-data"></a></dt>
<dd><p>logical; show (possibly normalized and scaled) data points?</p></dd>


<dt id="arg-show-fit-ci">show_fit_ci<a class="anchor" aria-label="anchor" href="#arg-show-fit-ci"></a></dt>
<dd><p>logical; show credible interval for the fitted trend?</p></dd>


<dt id="arg-show-data-ci">show_data_ci<a class="anchor" aria-label="anchor" href="#arg-show-data-ci"></a></dt>
<dd><p>logical; show bootstrapped quantile for data points?</p></dd>


<dt id="arg-ci-prob">ci_prob<a class="anchor" aria-label="anchor" href="#arg-ci-prob"></a></dt>
<dd><p>"se" for standard error bands (see <code><a href="https://rdrr.io/pkg/mgcv/man/predict.gam.html" class="external-link">mgcv::predict.gam()</a></code>), or numeric for simulation-based intervals.
If numerical, sets the probability for the simulation-based estimates of credible interval.</p></dd>


<dt id="arg-remove-null">remove_null<a class="anchor" aria-label="anchor" href="#arg-remove-null"></a></dt>
<dd><p>logical; only plot differentially expressed transcripts (not applicable for gene-only analyses)</p></dd>


<dt id="arg-null-threshold">null_threshold<a class="anchor" aria-label="anchor" href="#arg-null-threshold"></a></dt>
<dd><p>numeric; P value threshold for filtering out NULL transcripts</p></dd>


<dt id="arg-null-threshold-adj">null_threshold_adj<a class="anchor" aria-label="anchor" href="#arg-null-threshold-adj"></a></dt>
<dd><p>logical; use adjusted (default) or non-adjusted p-values for filtering targets</p></dd>


<dt id="arg-k-mult">k_mult<a class="anchor" aria-label="anchor" href="#arg-k-mult"></a></dt>
<dd><p>numerical; multiplier for the number of knots in the spline. Not recommended to change this value.</p></dd>


<dt id="arg-return-fits-only">return_fits_only<a class="anchor" aria-label="anchor" href="#arg-return-fits-only"></a></dt>
<dd><p>logical; return the model fits. Does not plot the function</p></dd>


<dt id="arg-family">family<a class="anchor" aria-label="anchor" href="#arg-family"></a></dt>
<dd><p>character; negative binomial ("nb", default) or Gaussian ("gaussian")</p></dd>


<dt id="arg-common-y-scale">common_y_scale<a class="anchor" aria-label="anchor" href="#arg-common-y-scale"></a></dt>
<dd><p>logical; for faceted plots of multiple transcripts, should the scale of the y-axis
be common or free.</p></dd>


<dt id="arg-scaled">scaled<a class="anchor" aria-label="anchor" href="#arg-scaled"></a></dt>
<dd><p>logical; scaled data by overdispersions (for bootstrapped data only)</p></dd>


<dt id="arg-base-size">base_size<a class="anchor" aria-label="anchor" href="#arg-base-size"></a></dt>
<dd><p>numeric; base font size for the plot</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ggplot object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Plots the fitted trend and data points for a given gene or target. If a gene ID
is supplied, the function will plot all transcripts for that gene.
The function can also be used to return the model fit(s) only, which are
<code>gamObject</code> objects from the <code>mgcv</code> package.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://l-a-yates.github.io/cpam/">cpam</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load gene-only example cpam object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"cpo_example.rda"</span>, package <span class="op">=</span> <span class="st">"cpam"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># example gene</span></span></span>
<span class="r-in"><span><span class="fu">plot_cpam</span><span class="op">(</span><span class="va">cpo_example</span>, gene_id <span class="op">=</span> <span class="st">"g003"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_cpam-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># gene with estimated changepoint at timepoint 3</span></span></span>
<span class="r-in"><span><span class="fu">plot_cpam</span><span class="op">(</span><span class="va">cpo_example</span>, gene_id <span class="op">=</span> <span class="st">"g013"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_cpam-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># manually set the changepoint</span></span></span>
<span class="r-in"><span><span class="fu">plot_cpam</span><span class="op">(</span><span class="va">cpo_example</span>, gene_id <span class="op">=</span> <span class="st">"g013"</span>, cp_fix <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="plot_cpam-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luke Yates, Michael Charleston, Jazmine Humphreys, Steven Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

