<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Prepare a cpam object — prepare_cpam • cpam</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prepare a cpam object — prepare_cpam"><meta name="description" content="Prepare a cpam object"><meta property="og:description" content="Prepare a cpam object"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cpam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cpam.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/l-a-yates/cpam/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Prepare a cpam object</h1>
      <small class="dont-index">Source: <a href="https://github.com/l-a-yates/cpam/blob/main/R/prepare_cpam.R" class="external-link"><code>R/prepare_cpam.R</code></a></small>
      <div class="d-none name"><code>prepare_cpam.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Prepare a cpam object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prepare_cpam</span><span class="op">(</span></span>
<span>  <span class="va">exp_design</span>,</span>
<span>  count_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  t2g <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  import_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  model_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"case-only"</span>, <span class="st">"case-control"</span><span class="op">)</span>,</span>
<span>  bootstrap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_fun <span class="op">=</span> <span class="st">"ts_filter"</span>,</span>
<span>  filter_fun_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min_reads <span class="op">=</span> <span class="fl">5</span>, min_prop <span class="op">=</span> <span class="fl">3</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span>,</span>
<span>  regularize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gene_level <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  aggregate_to_gene <span class="op">=</span> <span class="op">!</span><span class="va">gene_level</span>,</span>
<span>  condition_var <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>  case_value <span class="op">=</span> <span class="st">"treatment"</span>,</span>
<span>  num_cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fixed_effects <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  intercept_cc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="va">condition_var</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-exp-design">exp_design<a class="anchor" aria-label="anchor" href="#arg-exp-design"></a></dt>
<dd><p>a dataframe or tibble with the experimental design, containing at least a 'time' and a 'sample' column</p></dd>


<dt id="arg-count-matrix">count_matrix<a class="anchor" aria-label="anchor" href="#arg-count-matrix"></a></dt>
<dd><p>a matrix of counts. Column names must be in 'sample' column of <code>exp_design</code>,</p></dd>


<dt id="arg-t-g">t2g<a class="anchor" aria-label="anchor" href="#arg-t-g"></a></dt>
<dd><p>a transcript to gene dataframe or tibble with columns target_id and gene_id</p></dd>


<dt id="arg-import-type">import_type<a class="anchor" aria-label="anchor" href="#arg-import-type"></a></dt>
<dd><p>software used for quantification, one of "kallisto", "salmon" ,.... Ignored if <code>count_matrix</code> is supplied.</p></dd>


<dt id="arg-model-type">model_type<a class="anchor" aria-label="anchor" href="#arg-model-type"></a></dt>
<dd><p>"case-only" (default) or "case-control"</p></dd>


<dt id="arg-bootstrap">bootstrap<a class="anchor" aria-label="anchor" href="#arg-bootstrap"></a></dt>
<dd><p>logical; load bootstrap samples, also called inferential replicates, if available, and rescale counts.</p></dd>


<dt id="arg-filter-fun">filter_fun<a class="anchor" aria-label="anchor" href="#arg-filter-fun"></a></dt>
<dd><p>filter function to remove lowly expressed genes (default is <code>filter_fun()</code>)</p></dd>


<dt id="arg-filter-fun-args">filter_fun_args<a class="anchor" aria-label="anchor" href="#arg-filter-fun-args"></a></dt>
<dd><p>arguments for filter function</p></dd>


<dt id="arg-regularize">regularize<a class="anchor" aria-label="anchor" href="#arg-regularize"></a></dt>
<dd><p>logical; use empirical Bayes regularization of dispersions (default is TRUE)</p></dd>


<dt id="arg-gene-level">gene_level<a class="anchor" aria-label="anchor" href="#arg-gene-level"></a></dt>
<dd><p>logical; aggregate counts to gene level before data preparation and modelling (default is FALSE)</p></dd>


<dt id="arg-aggregate-to-gene">aggregate_to_gene<a class="anchor" aria-label="anchor" href="#arg-aggregate-to-gene"></a></dt>
<dd><p>logical; aggregate p values from transcript- to gene-level</p></dd>


<dt id="arg-condition-var">condition_var<a class="anchor" aria-label="anchor" href="#arg-condition-var"></a></dt>
<dd><p>string; column name in <code>exp_design</code> for the condition variable (for <code>model_type</code> = "case_control" only)</p></dd>


<dt id="arg-case-value">case_value<a class="anchor" aria-label="anchor" href="#arg-case-value"></a></dt>
<dd><p>value of <code>condition_var</code> that indicates the "case". All other values are deemed to be control</p></dd>


<dt id="arg-num-cores">num_cores<a class="anchor" aria-label="anchor" href="#arg-num-cores"></a></dt>
<dd><p>integer; number of cores to use for parallel computation</p></dd>


<dt id="arg-normalize">normalize<a class="anchor" aria-label="anchor" href="#arg-normalize"></a></dt>
<dd><p>logical; use model offsets based on sampling depth and gene length</p></dd>


<dt id="arg-fixed-effects">fixed_effects<a class="anchor" aria-label="anchor" href="#arg-fixed-effects"></a></dt>
<dd><p>a model formula of the form <code>~ effect1 + effect2</code></p></dd>


<dt id="arg-intercept-cc">intercept_cc<a class="anchor" aria-label="anchor" href="#arg-intercept-cc"></a></dt>
<dd><p>string; intercept for case-control model: "1" (default) for common intercept  or "condition"</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an object of class <code><a href="cpam-class.html">cpam-class</a></code>. The returned object has
methods <code>print</code> and <code>summary</code> for displaying information.
See <code><a href="cpam-class.html">cpam-class</a></code> for details on the structure of the returned object.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function prepares a cpam object for analysis.
The function loads count data from files or a matrix,
filters lowly expressed genes, computes normalisation factors,
and estimates dispersions. Many of these steps can be customised or turned off.</p>
<p>When bootstrap samples (inferential replicates) are available, it loads and
summarises these using means, standard errors, and estimated overdispersions.
The latter are a measure of quantification uncertainty and they are used to
rescale the counts which accounts for this uncertainty during the modelling steps.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Pedro L Baldoni, Yunshun Chen, Soroor Hediyeh-zadeh, Yang Liao, Xueyi Dong,
Matthew E Ritchie, Wei Shi, Gordon K Smyth,
Dividing out quantification uncertainty allows efficient assessment of
differential transcript expression with edgeR,
Nucleic Acids Research, Volume 52, Issue 3, 9 February 2024,
Page e13, https://doi.org/10.1093/nar/gkad1167</p>
<p>Yunshun Chen, Lizhong Chen, Aaron T L Lun, Pedro L Baldoni, Gordon K Smyth,
edgeR v4: powerful differential analysis of sequencing data with
expanded functionality and improved support for small counts
and larger datasets, Nucleic Acids Research, Volume 53, Issue 2,
27 January 2025, https://doi.org/10.1093/nar/gkaf018</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://l-a-yates.github.io/cpam/">cpam</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># gene-only example using example</span></span></span>
<span class="r-in"><span><span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu">prepare_cpam</span><span class="op">(</span>exp_design <span class="op">=</span> <span class="va">exp_design_example</span>,</span></span>
<span class="r-in"><span>                    count_matrix <span class="op">=</span> <span class="va">count_matrix_example</span>,</span></span>
<span class="r-in"><span>                    gene_level <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Processing count matrix</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Processing count matrix <span style="color: #B2B2B2;">[18ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Filtering low count genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Estimating dispersions using <span style="color: #0000BB;">edgeR</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Estimating dispersions using <span style="color: #0000BB;">edgeR</span> <span style="color: #B2B2B2;">[309ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Filtering low count genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Filtering low count genes <span style="color: #B2B2B2;">[457ms]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">cpo</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">cpam object</span> <span style="color: #00BBBB;">─────────────────────────────────────────────────────────────────</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="font-weight: 100;">case-only time series</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="font-weight: 100;">30 samples</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="font-weight: 100;">6 time points</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> • <span style="font-weight: 100;">Counts aggregated for gene-level inference</span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># transcript-level analysis with kallisto counts</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example Experimental Design</span></span></span>
<span class="r-in"><span><span class="va">exp_design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"s"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"path/"</span>,<span class="va">sample</span>,<span class="st">"/abundance.h5"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example Transcript-to-Gene Mapping</span></span></span>
<span class="r-in"><span><span class="va">t2g</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu">read_csv</span><span class="op">(</span><span class="st">"path/to/t2g.csv"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Prepare a cpam object</span></span></span>
<span class="r-in"><span><span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu">prepare_cpam</span><span class="op">(</span></span></span>
<span class="r-in"><span> exp_design <span class="op">=</span> <span class="va">exp_design</span>,</span></span>
<span class="r-in"><span> t2g <span class="op">=</span> <span class="va">t2g</span>,</span></span>
<span class="r-in"><span> import_type <span class="op">=</span> <span class="st">"kallisto"</span>,</span></span>
<span class="r-in"><span> num_cores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Print the object</span></span></span>
<span class="r-in"><span><span class="va">cpo</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luke Yates, Michael Charleston, Jazmine Humphreys, Steven Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

