<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>cpam • cpam</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="cpam">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cpam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/cpam.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/l-a-yates/cpam/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>cpam</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/l-a-yates/cpam/blob/main/vignettes/cpam.Rmd" class="external-link"><code>vignettes/cpam.Rmd</code></a></small>
      <div class="d-none name"><code>cpam.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="about">About<a class="anchor" aria-label="anchor" href="#about"></a>
</h3>
<p>This tutorial demonstrates the R package <a href="https://l-a-yates.github.io/cpam/"><code>cpam</code></a> for the
analysis of time series omics data. It serves as a basic introduction to
the package. There are also two detailed case studies using real world
data:</p>
<ul>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/torre.html" class="external-link">Arabidopsis
Case Study</a></li>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/crisp.html" class="external-link">Human
embryo Case Study</a></li>
</ul>
<p>These case studies and several simulation studies are presented in
the accompanying <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">manuscript</a> by <span class="citation">Yates et al. (2024)</span>. See also, the package <a href="https://l-a-yates.github.io/cpam/">website</a>.</p>
</div>
<div class="section level3">
<h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>The data for the following examples have been simulated based on
empirical RNA-seq data. These data are gene-level counts from a
case-only design with 6 time points and 5 replicates per time point.
Code to reproduce the data is available in this <a href="https://github.com/l-a-yates/cpam_manuscript" class="external-link">repository</a>.</p>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>You can install <a href="https://l-a-yates.github.io/cpam/"><code>cpam</code></a>
using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"cpam"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://l-a-yates.github.io/cpam/">cpam</a></span><span class="op">)</span>  </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p>An example experimental design is included in the <code>cpam</code>
package. Since it is case-only design, there are no experimental
conditions beyond time.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_design_example</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 30 × 2</span></span></span>
<span><span class="co">#&gt;    sample  time</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> X1         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> X2         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> X3         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> X4         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> X5         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> X6         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> X7         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> X8         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> X9         2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> X10        2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="count-matrix">Count matrix<a class="anchor" aria-label="anchor" href="#count-matrix"></a>
</h3>
<p>The example count data for are provided as a matrix. Let’s take a
look at the first few rows.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">count_matrix_example</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> </span>
<span><span class="co">#&gt;          X1     X2     X3     X4     X5     X6     X7     X8     X9    X10    X11    X12    X13    X14    X15    X16</span></span>
<span><span class="co">#&gt; g001 100702  93137 100748  90302  95962  79708  89726  85350  94282 121842 102158  90054 103692  78474  80961  81403</span></span>
<span><span class="co">#&gt; g002   8654   7926   8930   9196   8882   7993   8479   9865   8015   7060   9523   7124   9004   7840  10012   8161</span></span>
<span><span class="co">#&gt; g003   3258   3131   3024   3507   3687   2419   2456   2553   2238   2588   2250   2098   2166   1884   2339   1807</span></span>
<span><span class="co">#&gt; g004  24572  18575  28039  27230  25244  24315  26200  24790  26165  22652  24722  20735  27313  30438  26432  25533</span></span>
<span><span class="co">#&gt; g005 125565 118382 115517 105728 114457 120633 115978 108458 113169 115742 132714 121729  96647 134781 123007 124228</span></span>
<span><span class="co">#&gt; g006   1988   2058   2530   2135   2131   2139   2043   2068   2298   2281   2021   2097   2237   1951   1970   2175</span></span>
<span><span class="co">#&gt;         X17    X18   X19    X20    X21    X22    X23    X24    X25    X26    X27    X28    X29   X30</span></span>
<span><span class="co">#&gt; g001  76614  87686 81953 106238  87308  83021  81013  95695 109277  97975  79013  88479 119039 80521</span></span>
<span><span class="co">#&gt; g002   8789   9250  8816   7514   9565   7690   8946   7980   8919   9059   8227   7858   9253  9145</span></span>
<span><span class="co">#&gt; g003   1961   1791  1763   1893   1577   1505   1628   1511   1757   1573   1512   1659   1514  1355</span></span>
<span><span class="co">#&gt; g004  23079  27307 24375  26901  27090  28163  26881  26964  23837  24079  21432  23805  21358 29200</span></span>
<span><span class="co">#&gt; g005 126771 110148 95926 105201 100622 110298 145240 129146 114071 132796 125034 124525 115182 87204</span></span>
<span><span class="co">#&gt; g006   1990   2035  2295   2051   2159   1792   2003   2190   2140   2130   2281   2036   2241  2182</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fitting-cpam">Fitting <code>cpam</code><a class="anchor" aria-label="anchor" href="#fitting-cpam"></a>
</h3>
<p>To fit the models, we first prepare the <code>cpam</code> object,
then compute p-values, estimate changepoints, and select the shape for
each gene. In this simple example, simulated data are gene-level, that
is we do not have isoform-level counts. As such, we leave
```<code>the transcript-to-gene mapping (</code>t2g<code>) and we set</code>gene_level
= T`.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_cpam.html">prepare_cpam</a></span><span class="op">(</span>exp_design <span class="op">=</span> <span class="va">exp_design_example</span>,</span>
<span>                      count_matrix <span class="op">=</span> <span class="va">count_matrix_example</span>,</span>
<span>                      model_type <span class="op">=</span> <span class="st">"case-only"</span>,</span>
<span>                      t2g <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                      gene_level <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                      num_cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># just for the example</span></span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_p_values.html">compute_p_values</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="co"># 6 seconds</span></span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimate_changepoint.html">estimate_changepoint</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="co"># 4 seconds</span></span>
<span>  <span class="va">cpo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_shape.html">select_shape</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="co"># 5 seconds</span></span></code></pre></div>
<p>We can look at a summary of the fitted cpam object</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cpo</span></span></code></pre></div>
<p>If you run the code on your own computer, you can launch the Shiny
app to visualise the results interactively using
<code>visualise(cpo)</code>.</p>
</div>
<div class="section level3">
<h3 id="result-tables">Result tables<a class="anchor" aria-label="anchor" href="#result-tables"></a>
</h3>
<p>The results of the analysis are summarised using the
<code>results</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 104 × 16</span></span></span>
<span><span class="co">#&gt;    target_id         p    cp shape lfc.1  lfc.2  lfc.3  lfc.4  lfc.5  lfc.6 counts.1 counts.2 counts.3 counts.4 counts.5</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> g003      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 mdcx      0 -<span style="color: #BB0000;">0.349</span> -<span style="color: #BB0000;">0.634</span> -<span style="color: #BB0000;">0.856</span> -<span style="color: #BB0000;">1.02</span>  -<span style="color: #BB0000;">1.11</span>    <span style="text-decoration: underline;">3</span>267.    <span style="text-decoration: underline;">2</span>566.    <span style="text-decoration: underline;">2</span>105.    <span style="text-decoration: underline;">1</span>805.    <span style="text-decoration: underline;">1</span>617. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> g013      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0  0      0      0.327  0.655  0.982  <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">23</span>416.   <span style="text-decoration: underline;">29</span>382. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> g055      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 ilin      0  0.195  0.389  0.584  0.778  0.973    608.     696.     796.     911.    <span style="text-decoration: underline;">1</span>043. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> g063      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.595  0.937  1.03   0.861  0.443   <span style="text-decoration: underline;">6</span>172.    <span style="text-decoration: underline;">9</span>322.   <span style="text-decoration: underline;">11</span>814.   <span style="text-decoration: underline;">12</span>563.   <span style="text-decoration: underline;">11</span>210. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> g069      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     2 ilin      0  0      0.225  0.451  0.676  0.902  <span style="text-decoration: underline;">67</span>904.   <span style="text-decoration: underline;">67</span>904.   <span style="text-decoration: underline;">79</span>391.   <span style="text-decoration: underline;">92</span>821.  <span style="text-decoration: underline;">108</span>524. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> g090      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     2 ilin      0  0      0.239  0.477  0.716  0.954     32.8     32.8     38.7     45.7     53.9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> g106      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 mdcx      0 -<span style="color: #BB0000;">0.470</span> -<span style="color: #BB0000;">0.778</span> -<span style="color: #BB0000;">0.892</span> -<span style="color: #BB0000;">0.902</span> -<span style="color: #BB0000;">0.902</span>  <span style="text-decoration: underline;">30</span>076.   <span style="text-decoration: underline;">21</span>709.   <span style="text-decoration: underline;">17</span>542.   <span style="text-decoration: underline;">16</span>202.   <span style="text-decoration: underline;">16</span>090. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> g126      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 ilin      0  0.225  0.450  0.675  0.900  1.12    <span style="text-decoration: underline;">4</span>017.    <span style="text-decoration: underline;">4</span>695.    <span style="text-decoration: underline;">5</span>487.    <span style="text-decoration: underline;">6</span>412.    <span style="text-decoration: underline;">7</span>494. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> g128      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 ilin      0  0.209  0.418  0.628  0.837  1.05   <span style="text-decoration: underline;">38</span>099.   <span style="text-decoration: underline;">44</span>043.   <span style="text-decoration: underline;">50</span>915.   <span style="text-decoration: underline;">58</span>859.   <span style="text-decoration: underline;">68</span>043. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> g129      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.626  0.979  1.11   1.08   0.888    259.     399.     510.     560.     545. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 94 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: counts.6 &lt;dbl&gt;</span></span></span></code></pre></div>
<p>The generated results can be filtered by specifying minimum counts,
minimum log-fold changes, and maximum
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-values.
For example, to return only the transcripts with a log-fold change
greater than 1, at least 10 counts, and a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value
less than 0.01, we can run</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span>, min_count <span class="op">=</span> <span class="fl">10</span>, min_lfc <span class="op">=</span> <span class="fl">1</span>, p_threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 22 × 16</span></span></span>
<span><span class="co">#&gt;    target_id         p    cp shape lfc.1  lfc.2  lfc.3  lfc.4 lfc.5 lfc.6 counts.1 counts.2 counts.3 counts.4 counts.5</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> g063      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.595  0.937  1.03  0.861 0.443   <span style="text-decoration: underline;">6</span>172.    <span style="text-decoration: underline;">9</span>322.   <span style="text-decoration: underline;">11</span>814.   <span style="text-decoration: underline;">12</span>563.   <span style="text-decoration: underline;">11</span>210. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> g126      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 ilin      0  0.225  0.450  0.675 0.900 1.12    <span style="text-decoration: underline;">4</span>017.    <span style="text-decoration: underline;">4</span>695.    <span style="text-decoration: underline;">5</span>487.    <span style="text-decoration: underline;">6</span>412.    <span style="text-decoration: underline;">7</span>494. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> g128      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 ilin      0  0.209  0.418  0.628 0.837 1.05   <span style="text-decoration: underline;">38</span>099.   <span style="text-decoration: underline;">44</span>043.   <span style="text-decoration: underline;">50</span>915.   <span style="text-decoration: underline;">58</span>859.   <span style="text-decoration: underline;">68</span>043. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> g129      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.626  0.979  1.11  1.08  0.888    259.     399.     510.     560.     545. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> g171      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.593  0.922  1.01  0.901 0.616   <span style="text-decoration: underline;">4</span>950.    <span style="text-decoration: underline;">7</span>468.    <span style="text-decoration: underline;">9</span>382.    <span style="text-decoration: underline;">9</span>989.    <span style="text-decoration: underline;">9</span>241. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> g186      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cx        0 -<span style="color: #BB0000;">0.233</span> -<span style="color: #BB0000;">0.324</span> -<span style="color: #BB0000;">0.209</span> 0.219 1.05   <span style="text-decoration: underline;">19</span>740.   <span style="text-decoration: underline;">16</span>792.   <span style="text-decoration: underline;">15</span>774.   <span style="text-decoration: underline;">17</span>077.   <span style="text-decoration: underline;">22</span>974. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> g331      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0  0      0      0.339 0.678 1.02      37.0     37.0     37.0     46.8     59.1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> g334      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 micv      0  0.402  0.692  0.886 1.00  1.06    <span style="text-decoration: underline;">2</span>798.    <span style="text-decoration: underline;">3</span>696.    <span style="text-decoration: underline;">4</span>519.    <span style="text-decoration: underline;">5</span>171.    <span style="text-decoration: underline;">5</span>614. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> g341      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cx        0 -<span style="color: #BB0000;">0.465</span> -<span style="color: #BB0000;">0.593</span> -<span style="color: #BB0000;">0.383</span> 0.164 1.05   <span style="text-decoration: underline;">23</span>681.   <span style="text-decoration: underline;">17</span>157.   <span style="text-decoration: underline;">15</span>704.   <span style="text-decoration: underline;">18</span>161.   <span style="text-decoration: underline;">26</span>532. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> g393      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 cv        0  0.432  0.747  0.945 1.02  0.987    932.    <span style="text-decoration: underline;">1</span>257.    <span style="text-decoration: underline;">1</span>564.    <span style="text-decoration: underline;">1</span>793.    <span style="text-decoration: underline;">1</span>896. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 12 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: counts.6 &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-genes-and-transcripts">Plotting genes and transcripts<a class="anchor" aria-label="anchor" href="#plotting-genes-and-transcripts"></a>
</h3>
<p>A single gene can be plotted using the <code>plot_cpam</code>
function. Here we plot the gene g063</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpam.html">plot_cpam</a></span><span class="op">(</span><span class="va">cpo</span>, gene_id <span class="op">=</span> <span class="st">"g063"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/plot-g063-1.png" width="100%">
The subtitle shows <code>(1,cv)</code> indicating a changepoint at first
time point (i.e., the gene responds immediately) and a convex
(<code>'cv'</code>) shaped trend.</p>
<p>Let’s look for a gene that has a more complex trend. Unconstrained
shapes in <code>cpam</code> are denoted by
<code>'tp'</code>.<sup><a href="#fn1">1</a></sup> We can filter the
results for genes with unconstrained shapes and plot one of them.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">shape</span> <span class="op">==</span> <span class="st">"tp"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 14 × 16</span></span></span>
<span><span class="co">#&gt;    target_id         p    cp shape lfc.1  lfc.2  lfc.3   lfc.4  lfc.5   lfc.6 counts.1 counts.2 counts.3 counts.4</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> g210      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0  0.639  0.835  0.662   0.624  0.842    <span style="text-decoration: underline;">62</span>936.   <span style="text-decoration: underline;">98</span>024.  <span style="text-decoration: underline;">112</span>246.   <span style="text-decoration: underline;">99</span>582.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> g325      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0 -<span style="color: #BB0000;">0.897</span> -<span style="color: #BB0000;">0.882</span> -<span style="color: #BB0000;">0.157</span>   0.209 -<span style="color: #BB0000;">0.089</span><span style="color: #BB0000; text-decoration: underline;">1</span>   <span style="text-decoration: underline;">68</span>092.   <span style="text-decoration: underline;">36</span>565.   <span style="text-decoration: underline;">36</span>941.   <span style="text-decoration: underline;">61</span>074.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> g335      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0  0.252  0.311  0.240   0.454  0.980     <span style="text-decoration: underline;">1</span>154.    <span style="text-decoration: underline;">1</span>375.    <span style="text-decoration: underline;">1</span>432.    <span style="text-decoration: underline;">1</span>364.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> g339      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0 -<span style="color: #BB0000;">0.756</span> -<span style="color: #BB0000;">0.777</span> -<span style="color: #BB0000;">0.306</span>  -<span style="color: #BB0000;">0.368</span> -<span style="color: #BB0000;">0.914</span>     <span style="text-decoration: underline;">4</span>363.    <span style="text-decoration: underline;">2</span>584.    <span style="text-decoration: underline;">2</span>546.    <span style="text-decoration: underline;">3</span>529.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> g399      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0  0.248  0.487  0.705   0.875  1.01      <span style="text-decoration: underline;">3</span>916.    <span style="text-decoration: underline;">4</span>652.    <span style="text-decoration: underline;">5</span>487.    <span style="text-decoration: underline;">6</span>384.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> g400      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0 -<span style="color: #BB0000;">0.917</span> -<span style="color: #BB0000;">0.674</span>  0.558   0.706 -<span style="color: #BB0000;">0.270</span>     <span style="text-decoration: underline;">9</span>731.    <span style="text-decoration: underline;">5</span>155.    <span style="text-decoration: underline;">6</span>097.   <span style="text-decoration: underline;">14</span>330.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> g417      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0 -<span style="color: #BB0000;">0.425</span> -<span style="color: #BB0000;">0.574</span> -<span style="color: #BB0000;">0.531</span>  -<span style="color: #BB0000;">0.711</span> -<span style="color: #BB0000;">0.983</span>     <span style="text-decoration: underline;">7</span>235.    <span style="text-decoration: underline;">5</span>387.    <span style="text-decoration: underline;">4</span>859.    <span style="text-decoration: underline;">5</span>007.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> g424      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0 -<span style="color: #BB0000;">0.325</span> -<span style="color: #BB0000;">0.227</span>  0.044<span style="text-decoration: underline;">2</span> -<span style="color: #BB0000;">0.204</span> -<span style="color: #BB0000;">0.905</span>     <span style="text-decoration: underline;">9</span>559.    <span style="text-decoration: underline;">7</span>632.    <span style="text-decoration: underline;">8</span>170.    <span style="text-decoration: underline;">9</span>857.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> g640      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0  0.575  0.767  0.395   0.497  0.921     <span style="text-decoration: underline;">1</span>043.    <span style="text-decoration: underline;">1</span>554.    <span style="text-decoration: underline;">1</span>775.    <span style="text-decoration: underline;">1</span>372.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> g715      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     1 tp        0  0.836  0.810 -<span style="color: #BB0000;">0.239</span>  -<span style="color: #BB0000;">0.680</span> -<span style="color: #BB0000;">0.332</span>   <span style="text-decoration: underline;">157</span>101.  <span style="text-decoration: underline;">280</span>427.  <span style="text-decoration: underline;">275</span>440.  <span style="text-decoration: underline;">133</span>118.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: counts.5 &lt;dbl&gt;, counts.6 &lt;dbl&gt;</span></span></span></code></pre></div>
<p>We plot the first gene in the list.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpam.html">plot_cpam</a></span><span class="op">(</span><span class="va">cpo</span>, gene_id <span class="op">=</span> <span class="st">"g210"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/plot-g210-1.png" width="100%">
This selection of <code>'tp'</code> suggests that the trend for this
gene does not conform to one of the simpler shape types that
<code>cpam</code> uses. We can exclude <code>'tp'</code> as an option
and force <code>cpam</code> to choose among the simpler forms by setting
<code>shape_type = "shape2"</code> in the <code>plot_cpam</code>
function (<code>"shape1"</code>, the default, allows the
<code>'tp'</code>). This choice can be useful for analysis such as
clustering. For example:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpam.html">plot_cpam</a></span><span class="op">(</span><span class="va">cpo</span>, gene_id <span class="op">=</span> <span class="st">"g210"</span>,shape_type <span class="op">=</span> <span class="st">"shape2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/plot-g210-shape2-1.png" width="100%">
Here a monotonic increasing concave shape (‘micv’) is chosen, and we can
see this trend deviates from the data substantially more that the
unconstrained shape. See the <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">manuscript</a> for more
details on the shape types available in <code>cpam</code>.</p>
<p>Next we look for a gene with a changepoint by filtering for genes
with changepoints at the third time point.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">cp</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 13 × 16</span></span></span>
<span><span class="co">#&gt;    target_id         p    cp shape lfc.1 lfc.2 lfc.3  lfc.4  lfc.5  lfc.6 counts.1 counts.2 counts.3 counts.4 counts.5</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> g013      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0     0     0  0.327  0.655  0.982  <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">18</span>662.   <span style="text-decoration: underline;">23</span>416.   <span style="text-decoration: underline;">29</span>382. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> g146      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 dlin      0     0     0 -<span style="color: #BB0000;">0.329</span> -<span style="color: #BB0000;">0.659</span> -<span style="color: #BB0000;">0.988</span>    168.     168.     168.     134.     106. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> g187      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 dlin      0     0     0 -<span style="color: #BB0000;">0.419</span> -<span style="color: #BB0000;">0.839</span> -<span style="color: #BB0000;">1.26</span>   <span style="text-decoration: underline;">18</span>216.   <span style="text-decoration: underline;">18</span>216.   <span style="text-decoration: underline;">18</span>216.   <span style="text-decoration: underline;">13</span>622.   <span style="text-decoration: underline;">10</span>186. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> g297      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0     0     0  0.308  0.616  0.924   <span style="text-decoration: underline;">9</span>799.    <span style="text-decoration: underline;">9</span>799.    <span style="text-decoration: underline;">9</span>799.   <span style="text-decoration: underline;">12</span>133.   <span style="text-decoration: underline;">15</span>021. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> g304      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 cx        0     0     0 -<span style="color: #BB0000;">1.04</span>  -<span style="color: #BB0000;">0.652</span>  0.883   <span style="text-decoration: underline;">3</span>130.    <span style="text-decoration: underline;">3</span>130.    <span style="text-decoration: underline;">3</span>130.    <span style="text-decoration: underline;">1</span>520.    <span style="text-decoration: underline;">1</span>993. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> g331      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0     0     0  0.339  0.678  1.02      37.0     37.0     37.0     46.8     59.1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> g559      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0     0     0  0.292  0.584  0.877    444.     444.     444.     543.     665. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> g623      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 micv      0     0     0  0.317  0.634  0.951    329.     329.     329.     409.     510. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> g634      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 cx        0     0     0 -<span style="color: #BB0000;">0.880</span> -<span style="color: #BB0000;">1.13</span>  -<span style="color: #BB0000;">0.759</span>  <span style="text-decoration: underline;">12</span>547.   <span style="text-decoration: underline;">12</span>547.   <span style="text-decoration: underline;">12</span>547.    <span style="text-decoration: underline;">6</span>818.    <span style="text-decoration: underline;">5</span>721. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> g652      9.26<span style="color: #949494;">e</span><span style="color: #BB0000;">-319</span>     3 ilin      0     0     0  0.363  0.726  1.09   <span style="text-decoration: underline;">13</span>016.   <span style="text-decoration: underline;">13</span>016.   <span style="text-decoration: underline;">13</span>016.   <span style="text-decoration: underline;">16</span>739.   <span style="text-decoration: underline;">21</span>528. </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: counts.6 &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Again, we plot the first gene in the list.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cpam.html">plot_cpam</a></span><span class="op">(</span><span class="va">cpo</span>, gene_id <span class="op">=</span> <span class="st">"g013"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/plot-g013-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="clusters">Clusters<a class="anchor" aria-label="anchor" href="#clusters"></a>
</h3>
<p>The results function can be used to generate clusters according to
selected filters. The <code>plot_cluster</code> function can then be
used to visualise the clusters. With such a small simulated data set, we
don’t have many genes in each cluster, but we can try a few different
clustering options to get an idea of how the function works.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">cpo</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_cluster.html">plot_cluster</a></span><span class="op">(</span><span class="va">cpo</span>, <span class="va">res</span>, changepoints <span class="op">=</span> <span class="fl">1</span>, shapes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/clusters-1-1.png" width="100%">
There are 19 genes with a concave shape and a changepoint at the first
time point.</p>
<p>More than one shape or changepoint can be provided. For example:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_cluster.html">plot_cluster</a></span><span class="op">(</span><span class="va">cpo</span>, <span class="va">res</span>, changepoints <span class="op">=</span> <span class="fl">2</span>, shapes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dlin"</span>,<span class="st">"mdcx"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cpam_files/figure-html/clusters-2-1.png" width="100%">
There are just four genes with decreasing linear or monotonic decreasing
convex shapes which have a changepoint at the second time point.</p>
<p>Clustering can be further refined based on, for example, the rate at
which the above transcripts attain their maximum values. We illustrate
advanced refinements such as this <a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/torre.html" class="external-link">case
study</a>.</p>
</div>
<div class="section level3">
<h3 id="many-more-options">Many more options<a class="anchor" aria-label="anchor" href="#many-more-options"></a>
</h3>
<p>This is just a simple example to get you started. The package has
many more features and options. Check out the two case studies to see
how <code>cpam</code> can be used to analyse real-world data:</p>
<ul>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/torre.html" class="external-link">Arabidopsis
Case Study</a></li>
<li><a href="https://raw.githack.com/l-a-yates/cpam_manuscript/main/R/crisp.html" class="external-link">Human
Embryo Case Study</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><summary>
Click to expand
</summary><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [6] LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C           LC_TELEPHONE=C        </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.1 stringr_1.5.1 tidyr_1.3.1   dplyr_1.1.4   cpam_0.1.0   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] sass_0.4.9         utf8_1.2.4         generics_0.1.3     stringi_1.8.4      lattice_0.22-6     digest_0.6.37     </span></span>
<span><span class="co">#&gt;  [7] magrittr_2.0.3     RColorBrewer_1.1-3 evaluate_1.0.3     grid_4.4.3         fastmap_1.2.0      Matrix_1.7-2      </span></span>
<span><span class="co">#&gt; [13] jsonlite_1.9.0     limma_3.62.2       mgcv_1.9-1         purrr_1.0.4        scales_1.3.0       codetools_0.2-20  </span></span>
<span><span class="co">#&gt; [19] textshaping_1.0.0  jquerylib_0.1.4    cli_3.6.4          rlang_1.1.5        pbmcapply_1.5.1    munsell_0.5.1     </span></span>
<span><span class="co">#&gt; [25] splines_4.4.3      scam_1.2-18        withr_3.0.2        cachem_1.1.0       yaml_2.3.10        tools_4.4.3       </span></span>
<span><span class="co">#&gt; [31] parallel_4.4.3     colorspace_2.1-1   locfit_1.5-9.11    vctrs_0.6.5        R6_2.6.1           matrixStats_1.5.0 </span></span>
<span><span class="co">#&gt; [37] lifecycle_1.0.4    edgeR_4.4.2        fs_1.6.5           ragg_1.3.3         pkgconfig_2.0.3    desc_1.4.3        </span></span>
<span><span class="co">#&gt; [43] pkgdown_2.1.1      pillar_1.10.1      bslib_0.9.0        gtable_0.3.6       Rcpp_1.0.14        glue_1.8.0        </span></span>
<span><span class="co">#&gt; [49] statmod_1.5.0      systemfonts_1.2.1  xfun_0.51          tibble_3.2.1       tidyselect_1.2.1   knitr_1.49        </span></span>
<span><span class="co">#&gt; [55] farver_2.1.2       htmltools_0.5.8.1  nlme_3.1-167       labeling_0.4.3     rmarkdown_2.29     compiler_4.4.3    </span></span>
<span><span class="co">#&gt; [61] mvnfast_0.2.8</span></span></code></pre></div>
</details><br><div id="fn1">
<p>
<sup>1</sup><code>tp</code> stands for <em>thinplate</em> which is the
type of spline used for the ‘unconstrained’ curves as defined in the
<code>mgcv</code> package. The curves are still penalised to be smooth,
but the shape type is not fixed.
</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Yates2024" class="csl-entry">
Yates, Luke A., Jazmine L. Humphreys, Michael A. Charleston, and Steven
M. Smith. 2024. <span>“Shape-Constrained, Changepoint Additive Models
for Time Series Omics Data with <em>Cpam</em>.”</span> <em>BioRxiv</em>,
December. <a href="https://doi.org/10.1101/2024.12.22.630003" class="external-link">https://doi.org/10.1101/2024.12.22.630003</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Luke Yates, Michael Charleston, Jazmine Humphreys, Steven Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
